core/src/main/java/org/infinispan/statetransfer/StateTransferInterceptor.java

```diff
@@ -192,17 +186,26 @@ private InvocationStage visitReadCommand(InvocationContext ctx, AbstractTopology
          if (!(ce instanceof OutdatedTopologyException) && !(ce instanceof SuspectException))
             throw t;
 
-         // We increment the topology id so that updateTopologyIdAndWaitForTransactionData waits for the next topology.
+         // We increment the topology to waits for the next topology.
          // Without this, we could retry the command too fast and we could get the OutdatedTopologyException again.
+         int currentTopologyId = currentTopologyId();
          if (trace)
             log.tracef("Retrying command because of topology change, current topology is %d: %s",
-                  currentTopologyId(), rCommand);
+                  currentTopologyId, rCommand);
          AbstractTopologyAffectedCommand cmd = (AbstractTopologyAffectedCommand) rCommand;
-         int newTopologyId = Math.max(currentTopologyId(), cmd.getTopologyId() + 1);
+         int desiredTopologyId = cmd.getTopologyId() + 1;
+         if (ce instanceof OutdatedTopologyException) {
+            OutdatedTopologyException ote = (OutdatedTopologyException) ce;
+            if (ote.desiredTopologyId >= 0) {
+               desiredTopologyId = ote.desiredTopologyId;
+            }
+         }
+         int newTopologyId = Math.max(currentTopologyId, desiredTopologyId);
```

Comment:
Good catch. Thinking about names again, I'll probably rename that to `requestedTopologyId` - we shouldn't be too passionate about topologies :)